### PR TITLE
Introduce budget, reimbursements for expenses

### PR DESCRIPTION
**Pull request by raucao**
This is a big one. It isn't linked from the dashboard yet, but you can browse to `/budget` and propose reimbursements from there.

Requires https://github.com/67P/kredits-contracts/pull/198
Refs 67P/kredits-contracts#48

## To do

* [x] Fix ember linting errors
* [x] Sync new and missing reimbursements from network when cached data loaded
* [x] Fix cache loading after adding new reimbursements (records shows up twice)
* [x] Fix bug, which claims expense amounts are missing, after tx creation

### CODE REVIEW

**Comment by raucao** [comment]
I just finished the functionality of the PR, so there are just the linter errors left. You can test it on your machines now.

There's one known issue, which won't be much of a problem in a production: a race condition, when your tx confirms within the loading of the reimbursement list after pressing the submit button on the reimbursement form. I'm not sure how to handle that, and it doesn't break anything in the unlikely case that you see this  in production. Just reload the page and all is good, due to the caching and syncing logic.

**Comment by raucao** [comment]
All done and ready to merge from my point of view.

**Review by galfert — CHANGES_REQUESTED** [review]
Wow, that was a lot of changes :)

Left a few comments.

So far I wasn't able to test it manually, because of problems running the contracts (see XMPP).

File: app/components/add-reimbursement/component.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    const contributor = this.contributors.findBy('id', this.contributorId);
+
+    const attributes = {
+      amount: parseInt(parseFloat(this.total) * 100000000), // convert to sats
+      token: config.tokens['WBTC'],
+      contributorId: parseInt(this.contributorId),
+      title: `Expenses covered by ${contributor.name}`,
+      description: this.description,
+      url: this.url,
+      expenses: JSON.parse(JSON.stringify((this.expenses)))
```
[the remark this answers is not in the document]
Comment (raucao) [reply app/components/add-reimbursement/component.js:130]: It's to get rid of the EmberObject properties and only retain the pure expense data in the object handed to the wrapper.

File: app/services/exchange-rates.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import config from 'kredits-web/config/environment';
+
+// Need to go through proxy for CORS headers
+const bitstampBaseUrl = `${config.corsProxy}https://www.bitstamp.net/api/v2`;
+
+async function fetchFromBitstamp(currencyPair) {
+  try {
+    const res = await fetch(`${bitstampBaseUrl}/ticker/${currencyPair}/`).then(r => r.json());
+    return res.vwap; // Last 24 hours volume weighted average price
+  } catch(e) {
```
[the remark this answers is not in the document]
Comment (raucao) [reply app/services/exchange-rates.js:12]: If it is not OK, then this already just gracefully fails by not automatically converting numbers.

File: app/services/community-funds.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  @tracked balancesLoaded = false;
+  @tracked balances = A([]);
+
+  @task
+  *fetchBalances () {
+    const uri = `${txServiceBaseUrl}/balances/usd/`;
+
+    yield fetch(uri).then(res => res.json())
+      .then(res => this.processBalances(res))
+      .catch(err => {
```
[the remark this answers is not in the document]
Comment (raucao) [reply app/services/community-funds.js:19]: This gracefully fails by catching the exception in `processBalances` I think.

File: app/services/kredits.js [hunk 4]
```diff
@@ -406,23 +447,186 @@ export default Service.extend({
...
-    return this.kredits.Contributor
-      .functions.getContributorIdByAddress(this.currentUserAccounts.firstObject)
-      .then((id) => {
-        // check if the user is a contributor or not
-        if (id === 0) {
-          return Promise.resolve();
-        } else {
-          return this.kredits.Contributor.getById(id);
+    console.debug(`[kredits] Cached ${this[collection].length} ${collection} in browser storage`);
+    return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (raucao) [reply app/services/kredits.js:485]: Didn't know that. Not sure I like it being implicit here...

File: app/components/add-expense-item/component.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  get isValidAmount () {
+    return isValidAmount(this.amount);
+  }
+
+  get amountInputClass () {
+    return this.isValidTotal ? 'valid' : '';
+  }
+
+  get submitButtonEnabled () {
+    return true;
```
[the remark this answers is not in the document]
Comment (raucao) [reply app/components/add-expense-item/component.js:33]: I guess it's just a leftover from before the proper form validation.

**Comment by raucao** [comment]
@galfert Did you see my replies? I just merged the corresponding contracts PR, so it'd be great if we can merge this one as well now.

**Review by galfert — APPROVED** [review]
LGTM

**Pull request merged**